mllib/src/main/scala/org/apache/spark/mllib/clustering/DpMeans.scala

```diff
@@ -0,0 +1,247 @@
+ * local cluster center is found. A data point is said to be "covered" by
+ * a cluster `c` if the distance from the point to the cluster center of `c`
+ * is less than a given lambda value.
+ *
+ * The original paper is "MLbase: Distributed Machine Learning Made Easy" by
+ * Xinghao Pan, Evan R. Sparks, Andre Wibisono
+ *
+ * @param lambda The distance threshold value that controls cluster creation.
+ * @param convergenceTol The threshold value at which convergence is considered to have occurred.
+ * @param maxIterations The maximum number of iterations to perform.
+ */
+
+@Experimental
+class DpMeans private (
+    private var lambda: Double,
+    private var convergenceTol: Double,
+    private var maxIterations: Int) extends Serializable with Logging {
+
+  /**
+   * Constructs a default instance.The default parameters are {lambda: 1, convergenceTol: 0.01,
+   * maxIterations: 20}.
+   */
+  def this() = this(1, 0.01, 20)
+
+  /** Set the distance threshold that controls cluster creation. Default: 1 */
+  def getLambda(): Double = lambda
+
+  /** Return the lambda. */
+  def setLambda(lambda: Double): this.type = {
+    this.lambda = lambda
+    this
+  }
+
+  /** Set the threshold value at which convergence is considered to have occurred. Default: 0.01 */
+  def setConvergenceTol(convergenceTol: Double): this.type = {
+    this.convergenceTol = convergenceTol
+    this
+  }
+
+  /** Return the threshold value at which convergence is considered to have occurred. */
+  def getConvergenceTol: Double = convergenceTol
+
+  /** Set the maximum number of iterations. Default: 20 */
+  def setMaxIterations(maxIterations: Int): this.type = {
+    this.maxIterations = maxIterations
+    this
+  }
+
+  /** Return the maximum number of iterations. */
+  def getMaxIterations: Int = maxIterations
+
+  /**
+   * Perform DP means clustering
+   */
+  def run(data: RDD[Vector]): DpMeansModel = {
+    if (data.getStorageLevel == StorageLevel.NONE) {
+      logWarning("The input data is not directly cached, which may hurt performance if its"
+        + " parent RDDs are also uncached.")
+    }
+
+    // Compute squared norms and cache them.
+    val norms = data.map(Vectors.norm(_, 2.0))
+    norms.persist()
+    val zippedData = data.zip(norms).map {
+      case (v, norm) => new VectorWithNorm(v, norm)
+    }
+
+    // Implementation of DP means algorithm.
+    var iteration = 0
+    var covered = false
+    var converged = false
+    var localCenters = Array.empty[VectorWithNorm]
+    val globalCenters = ArrayBuffer.empty[VectorWithNorm]
+
+    // Execute clustering until the maximum number of iterations is reached
+    // or the cluster centers have converged.
+    while (iteration < maxIterations && !converged) {
+      type WeightedPoint = (Vector, Long)
+      def mergeClusters(x: WeightedPoint, y: WeightedPoint): WeightedPoint = {
+        axpy(1.0, x._1, y._1)
+        (y._1, x._2 + y._2)
```

Comment:
For merge operations, the contract used in Spark is that the first argument is mutable but not the second. So we should add `y` to `x` and return `x`.

It might be better to move this method near `reduceByKey`. See my comments about `reduceByKey`.